### PR TITLE
[Core][Node Labels 1/n] Add --labels param in ray start command to setting node labels

### PR DESCRIPTION
**Pull request by larrylian**
## Why are these changes needed?
1. Add --labels param in ray start command-line to setting node labels
```
ray start --head --labels='{"gpu_type": "A100", "region": "us"}'
```

2. Add a new field "Labels" to the ray.nodes() API to getting the labels of each node.
```
ray.nodes()[0]["Labels"] == {"gpu_type": "A100", "region": "us"}
```

3. Add a basic test case  
python/ray/tests/test_node_labels.py  
Other exceptional scenarios and boundary scenarios testing will be added later by new PR.

## Related issue number
Enhancing node affinity scheduling feature through node labels #34894
> (P1)Finalize the command-line API(ray start) for setting node labels.  

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/scripts/scripts.py [hunk 1]
```diff
@@ -524,6 +524,13 @@ def debug(address):
     default=False,
     help="If True, the usage stats collection will be disabled.",
 )
+@click.option(
+    "--labels",
+    required=False,
+    default="{}",
+    type=str,
+    help="Define the key-value format of node labels, which is a serialized JSON.",
```
Comment (jjyao) [inline python/ray/scripts/scripts.py:532]: a JSON serialized dictionary mapping label name to label value

File: python/ray/scripts/scripts.py [hunk 2]
```diff
@@ -524,6 +524,13 @@ def debug(address):
     default=False,
     help="If True, the usage stats collection will be disabled.",
 )
+@click.option(
+    "--labels",
```
Comment (jjyao) [inline python/ray/scripts/scripts.py:528]: maybe `hidden=True,` for now and we can make it public after we finish everything

File: python/ray/tests/test_node_labels.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import os
+import sys
+import pytest
+
+import ray
+
+
+@pytest.mark.parametrize(
+    "call_ray_start",
+    ['ray start --head --labels={"gpu_type":"A100","region":"us"}'],
```
Comment (jjyao) [inline python/ray/tests/test_node_labels.py:10]: can you also test edge cases like the label key and value contain whitespace etc

File: src/ray/raylet/main.cc [hunk 4]
```diff
@@ -74,8 +77,36 @@ DEFINE_string(plasma_directory,
...
+  if (labels_json_str.empty()) {
+    return labels;
+  }
+  try {
+    json j = json::parse(labels_json_str);
+    for (auto &el : j.items()) {
+      if (el.value().is_string()) {
+        labels.emplace(el.key(), el.value());
+      } else {
+        throw std::invalid_argument(
```
Comment (jjyao) [inline src/ray/raylet/main.cc:98]: we can just `RAY_LOG(FATAL)`

File: src/ray/raylet/main.cc [hunk 5]
```diff
@@ -74,8 +77,36 @@ DEFINE_string(plasma_directory,
...
+            "The value of the '" + el.key() +
+            "' field in the node labels configuration is not string type.");
+      }
+    }
+  } catch (const std::exception &e) {
+    RAY_LOG(ERROR) << "Failed to parse node labels json string for " << e.what();
```
Comment (jjyao) [inline src/ray/raylet/main.cc:104]: FATAL

**Review by jjyao — COMMENTED** [review]
Nice!!!

File: python/ray/tests/test_node_labels.py (hunk 3 above)
Comment (larrylian) [reply python/ray/tests/test_node_labels.py:10]: 1. I add  a new 'test_ray_start_set_empty_node_labels' edge cases.
2. I will add more test cases for parameter exceptions in the next PR, like the case which the label key and value contain whitespace

**Review by jjyao — APPROVED** [review]

**Comment by jjyao** [comment]
@pcmoritz could you review the core API change?

**Comment by ollie-iterators** [comment]
The linkcheck error has been fixed elsewhere in this commit: https://github.com/ray-project/ray/commit/d0153a394a2f9078a1861e48e764774657464f95

File: python/ray/_private/utils.py [hunk 6]
```diff
@@ -1896,3 +1896,32 @@ def update_envs(env_vars: Dict[str, str]):
             os.environ[key] = value.replace("${" + key + "}", os.environ.get(key, ""))
         else:
             os.environ[key] = value
+
+
+def parse_node_labels_json(
+    labels_json: str, cli_logger, cf, command_arg="--labels"
+) -> Dict[str, str]:
+    if not labels_json:
+        return {}
```
Comment (jiwq) [inline python/ray/_private/utils.py:1905]: The logical maybe should keep align with `parse_resources_json`, don't assign empty dict if the labels_json is null.

Comment (larrylian) [reply python/ray/_private/utils.py:1905]: No problem,  I has fixed.

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**